runtime/src/fees.rs

```diff
@@ -0,0 +1,242 @@
+			// hacky, without enough balance to keep the account alive, deposit_into_existing will no-op.
+			pallet_balances::GenesisConfig::<Runtime> {
+				balances: vec![(generate_account("Somebody"), ExistentialDeposit::get())],
+			}
+			.assimilate_storage(&mut t)
+			.unwrap();
+			t.into()
+		}
+	}
+	fn correct_and_deposit_fee_with_passing_defaults(
+		who: &AccountId32,
+		already_withdrawn: u128,
+		corrected_fee: u128,
+	) {
+		let dispatch_info = DispatchInfo::default();
+		let post_info = PostDispatchInfo::default();
+		let tip = Default::default();
+		let already_withdrawn = Some(NegativeImbalance::new(already_withdrawn));
+
+		<<Runtime as TxPaymentConfig>::OnChargeTransaction as OnChargeTransaction<
+					Runtime,
+				>>::correct_and_deposit_fee(
+					who,
+					&dispatch_info,
+					&post_info,
+					corrected_fee,
+					tip,
+					already_withdrawn,
+				)
+				.unwrap();
+	}
+
+	#[test]
+	fn year_offset() {
+		let test = || {
+			let f = |i: u32| {
+				CurrencyFeeRedemptionAdapter::<pallet_balances::Pallet<Runtime>,()>::bucketed_year_offset::<Runtime>(&i)
+			};
+			assert_eq!(BLOCK_YEAR, f(0));
+			assert_eq!(BLOCK_YEAR, f(9));
+			assert_eq!(BLOCK_YEAR + 10, f(10));
+			assert_eq!(BLOCK_YEAR + 20, f(20));
+			assert_eq!(BLOCK_YEAR + 90, f(99));
+			assert_eq!(BLOCK_YEAR + 100, f(100));
+		};
+		ExtBuilder::build().execute_with(test);
+	}
+
+	#[test]
+	fn fee_adapter_works() {
+		let test = || {
+			let acc = generate_account("Somebody");
+
+			//offset fees by a block-year
+			let year_offset = CurrencyFeeRedemptionAdapter::<pallet_balances::Pallet<Runtime>,()>::bucketed_year_offset::<Runtime>(&0);
+
+			// Adapter no-ops if adjusted_fees is 0
+			correct_and_deposit_fee_with_passing_defaults(&acc, 2, 0);
+			assert!(
+				pallet_creditcoin::RetainedFees::<Runtime>::get(&year_offset, &acc).is_none(),
+				"0-fees are no-ops"
+			);
+
+			//Adapter inserts if fees > 0
+			correct_and_deposit_fee_with_passing_defaults(&acc, 2, 1);
+			let fee = pallet_creditcoin::RetainedFees::<Runtime>::get(&year_offset, &acc)
+				.expect("fees inserted for redemption");
+			assert_eq!(fee, 1u128);
+
+			//Adapter aggregates
+			correct_and_deposit_fee_with_passing_defaults(&acc, 2, 1);
+			let fee = pallet_creditcoin::RetainedFees::<Runtime>::get(&year_offset, &acc)
+				.expect("fees inserted for redemption");
+			assert_eq!(fee, 2u128);
+		};
+
+		ExtBuilder::build().execute_with(test);
+	}
+
+	#[test]
+	fn creditcoin_on_init_redeems() {
+		let test = || {
+			let acc = generate_account("Somebody");
+			System::set_block_number(1);
+			//offset fees by a block-year
+			let year_offset = CurrencyFeeRedemptionAdapter::<pallet_balances::Pallet<Runtime>,()>::bucketed_year_offset::<Runtime>(&0);
+
+			correct_and_deposit_fee_with_passing_defaults(&acc, 2, 1);
+
+			let free = Balances::free_balance(&acc);
+			assert_eq!(free, existential_deposit() + 1u128);
+			//Give back excess fees worth one.
```

Comment:
I think a more realistic simulation will be if you call `Creditcoin:on_initialize()` for each block 1..year_offset. Assert that ballance had not changed (same as line 223).

Reply:
feel free to `roll_to` or `roll_by` to the desired block. Note: `roll_by` executes first before rolling.